GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,159 @@
+
+        // #define EAGAIN      35      /* Resource temporarily unavailable */
+        // #define EWOULDBLOCK EAGAIN  /* Operation would block */
+        private const int EWouldBlock = 35; 
+
+        private const int InvalidFileDescriptor = -1;
+
+        private static readonly ushort FileMode644 = Convert.ToUInt16("644", 8);
+
+        private int lockFileDescriptor;
+
+        public MacFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath,
+            string signature)
+            : base(fileSystem, tracer, lockPath, signature) 
+        {
+            this.lockFileDescriptor = InvalidFileDescriptor;
+        }
+
+        [Flags]
+        private enum FLockOperations
+        {
+            LockSh = 1, // #define LOCK_SH   1    /* shared lock */
+            LockEx = 2, // #define LOCK_EX   2    /* exclusive lock */
+            LockNb = 4, // #define LOCK_NB   4    /* don't block when locking */
+            LockUn = 8  // #define LOCK_UN   8    /* unlock */
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = NativeMethods.Open(this.LockPath, OpenCreate | OpenWriteOnly, FileMode644);
+                if (this.lockFileDescriptor == InvalidFileDescriptor)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Failed to open lock file");
+                    
+                    return false;
+                }
+            }
+
+            if (NativeMethods.FLock(this.lockFileDescriptor, (int)(FLockOperations.LockEx | FLockOperations.LockNb)) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+                if (errno != EIntr && errno != EWouldBlock)
+                {
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Unexpected error when locking file");
+                }
+
+                return false;
+            }
+
+            if (NativeMethods.FTruncate(this.lockFileDescriptor, 0) == 0)
+            {
+                byte[] signatureBytes = Encoding.UTF8.GetBytes(this.Signature);
+                long bytesWritten = NativeMethods.Write(
+                    this.lockFileDescriptor,
+                    signatureBytes,
+                    Convert.ToUInt64(signatureBytes.Length));
+
+                if (bytesWritten == -1)
+                {
+                    int errno = Marshal.GetLastWin32Error();

```

Comment:
Why are we still returning true if we hit this, or the other error right below it? Should we fail here?

Reply:
> Why are we still returning true if we hit this

We're returning true because `FLock` successfully acquired the lock

> or the other error right below it? Should we fail here?

I don't think that we need to.  Currently the signature is only to help with diagnostics, it need not be written for the locking to work correctly.  (I also considered not writing the signature at all, it would speed things up and eliminate a class of possible failures, thoughts on that option?)

Reply:
Yea that's a good point. The reason we added the signature in the first place is that we were also creating the index.lock from the mount process. It was useful to know, for diagnostics, if an abandoned index.lock had been created by git or by us. Now that we no longer create the index.lock... is the signature adding any value at all? Even on Windows? We could just drop it if it's not serving a purpose any longer.